### PR TITLE
[POC] Add index_select(lookup_tv, dim, index_tv) fusion for nvfuser

### PR DESCRIPTION
**Pull request by chang-l**
Feiwen:
1) support arbitrarily rank; 2) support arbitrarily dim value; 3) index_tv can become consumer; 4) lookup_tv cannot become consumer. TODO: move index calculation from codegen.cpp to lower_index.cpp

Chang: 
I cherry-picked this PR from this commit: https://github.com/azazhu/pytorch/commit/b1f7469f3f79574e0e7a993c5da032f27000920c with the following modification to make it compatible with latest nvfuser/devel branch:
- Added `shallow_copy()` and its implementation to IndexSelectOp class definition
- Explicitly converted to int64_t for the variable: `EvaluatorValue::prev_val` at [here](https://github.com/chang-l/pytorch-nvfuser/blob/0c6a43df160fe46a1cc7f30cc57d0150c31c418e/torch/csrc/jit/codegen/cuda/executor_utils.cpp#L847).
- Changed Feiwen's debug output in lower2device.cpp to make it compatible with devel branch
- Some header file changed in `test/cpp/jit/test_index_select.cpp` to compile.

## Tests
- Integrated unit tests `NVFuserTest.FusionIndexSelect*`  passed

@azazhu
I believe this initial PR, which rebases your work to nvfuser/TOT, is completed and passed all your unit tests. Maybe you want to refer this PR to avoid any repetitive work when creating your new PRs on top of latest devel branch.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -34,6 +34,19 @@ TensorView* maybe_broadcast_inner_to_rank(TensorView* t, size_t rank) {
   return t;
 }
 
+TensorView* maybe_broadcast_outter_to_rank(TensorView* t, size_t rank) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:37]: Can't we just use `std::vector<Val*> maybeBroadcast(const std::vector<Val*>& vals)` instead?

File: test/test_jit_nvfuser_graph.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import torch
```
Comment (jjsjann123) [inline test/test_jit_nvfuser_graph.py:1]: Note, python test should go into https://github.com/pytorch/pytorch/blob/master/test/test_jit_cuda_fuser.py

File: test/cpp/jit/test_index_select.cpp [hunk 3]
```diff
@@ -0,0 +1,722 @@
+#if defined(USE_CUDA)
```
Comment (naoyam) [inline test/cpp/jit/test_index_select.cpp:1]: Please move this to `torch/csrc/jit/codegen/cuda/test`.

File: test/test_jit_nvfuser_graph.py (hunk 2 above)
Comment (naoyam) [inline test/test_jit_nvfuser_graph.py:1]: Please merge this with the existing Python test (`test/test_jit_cuda_fuser.py`)

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 4]
```diff
@@ -226,15 +225,40 @@ void GpuLower::lower(Fusion* fusion, DataType index_type) {
 
   FusionGuard fg(fusion_);
   // prepare for lowering
+  if (isDebugDumpEnabled(DebugDumpOption::PassTransform)) {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp:228]: Please pull this feature out as a separate PR

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 5]
```diff
@@ -1514,6 +1514,36 @@ class IrParser {
       }
     }
 
+    {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1517]: Please pull the frontend change into a separate PR.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/arith.cpp:37]: no, as we have to broadcast the tensorview(the third arg of index_select) on inner dim and outter dim.

File: torch/csrc/jit/codegen/cuda/lower2device.cpp (hunk 4 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/lower2device.cpp:228]: Done.
debug_lower_GPU PR: https://github.com/azazhu/pytorch/tree/fw/dump_debug
frontend PR: https://github.com/azazhu/pytorch/tree/fw/idx_sel_frontend
backend PR: https://github.com/azazhu/pytorch/tree/fw/idx_sel_backend

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 5 above)
Comment (azazhu) [reply torch/csrc/jit/codegen/cuda/parser.cpp:1517]: Done.
debug_lower_GPU PR: https://github.com/azazhu/pytorch/tree/fw/dump_debug
frontend PR: https://github.com/azazhu/pytorch/tree/fw/idx_sel_frontend
backend PR: https://github.com/azazhu/pytorch/tree/fw/idx_sel_backend

File: test/cpp/jit/test_index_select.cpp (hunk 3 above)
Comment (azazhu) [reply test/cpp/jit/test_index_select.cpp:1]: done. 
see backend PR: https://github.com/azazhu/pytorch/tree/fw/idx_sel_backend

File: test/test_jit_nvfuser_graph.py (hunk 2 above)
Comment (azazhu) [reply test/test_jit_nvfuser_graph.py:1]: Jie, move to https://github.com/csarofeen/pytorch/blob/devel/test/test_jit_cuda_fuser.py?

**Comment by naoyam** [comment]
@azazhu Thanks for separating this into the three PRs. Please close this PR if everything is moved to the new PRs.
